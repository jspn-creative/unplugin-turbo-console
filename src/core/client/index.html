<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2/css/pico.min.css">
  <link rel="icon" type="image/svg+xml" href="https://unplugin.unjs.io/logo.svg">
  <title>Launch Editor</title>
</head>
<style>
  @media (prefers-color-scheme: dark) {
    body {
      background-color: black;
    }
  }
</style>

<body>
  <main class="container">
    <article>
      <h4>Ops, It get some error when launch to the editor:</h4>
      <div id="error-info" style="color: #EE402E">
        
      </div>
        <small>You can report this issus at <a href="https://github.com/unplugin/unplugin-turbo-console/issues"
            target="_blank">GitHub</a>.
        </small>
    </article>
  </main>
</body>
<script>
  ; (async () => {
    try {
      const position = window.location.hash.slice(1)
      const raw = await fetch(`/__open-in-editor?position=${(position)}`)
      const response = await raw.json()
      if (response.status === 'error')
        throw new Error(response.message)

      window.close()
    }
    catch (error) {
      const errorInfo = document.getElementById('error-info')
      errorInfo.innerText = String(error)
    }
  })()
</script>

</html>
