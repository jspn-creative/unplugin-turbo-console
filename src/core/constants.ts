export const PLUGIN_NAME = 'unplugin-turbo-console'
export const NUXT_CONFIG_KEY = 'turboConsole'
export const virtualModuleId = `~console`
export const resolvedVirtualModuleId = `\0${virtualModuleId}`
