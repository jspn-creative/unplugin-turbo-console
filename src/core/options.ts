import { env } from 'node:process'
import type { Options } from '../types'

const DETAULT_OPTIONS: Options = {
  prefix: '',
  suffix: '',
  disableLaunchEditor: false,
  disableHighlight: false,
  extendedPathFileNames: [],
  babelParserPlugins: ['typescript', 'jsx'],
  port: 3070,
  passServerLogs: true,
}

export const BUILD_OPTIONS: Options = {
  disableLaunchEditor: true,
  disableHighlight: true,
  passServerLogs: false,
}

export function resolveOptions(options: Options): Options {
  // merge babel parser plugins
  if (options.babelParserPlugins) {
    options.babelParserPlugins = [
      ...DETAULT_OPTIONS.babelParserPlugins!,
      ...options.babelParserPlugins,
    ]
  }

  let resolved = {
    ...DETAULT_OPTIONS,
    ...options,
  }

  if (env.NODE_ENV === 'production') {
    resolved = {
      ...resolved,
      ...BUILD_OPTIONS,
    }
  }

  return resolved
}
