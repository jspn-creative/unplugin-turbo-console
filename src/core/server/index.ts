import { createServer } from 'node:http'
import { cwd } from 'node:process'
import { readFile, stat } from 'node:fs/promises'
import { join, resolve } from 'pathe'
import { createApp, eventHandler, getQuery, serveStatic, toNodeListener } from 'h3'

// @ts-expect-error any
import launch from 'launch-editor'
import { CLIENT_DIR } from '../dir'

export async function startServer(port: number = 3070) {
  try {
    await fetch(`http://localhost:${port}/health`)
  }
  catch (error) {
    const app = createApp()

    app.use('/health', eventHandler(() => {
      return {
        message: 'ok',
      }
    }))

    app.use('/routeMap', eventHandler(() => {
      const routeMap = globalThis.TurboConsoleRouteMap || new Map()
      return Object.fromEntries(routeMap)
    }))

    app.use('/__open-in-editor', eventHandler(async (event) => {
      try {
        const { position } = getQuery(event) as { position: string }
        const routeMap = globalThis.TurboConsoleRouteMap

        if (!routeMap)
          throw new Error('routeMap is undefined')

        const parsed = position.split(',')
        const routeMapString = parsed[0]
        const line = parsed[1] || 1
        const column = parsed[2] || 1

        let filePath = ''

        for (const [key, value] of routeMap) {
          if (value === routeMapString) {
            filePath = key
            break
          }
        }

        launch(resolve(cwd(), `${filePath}:${line}:${column}`))
        return {
          status: 'success',
        }
      }
      catch (error) {
        return {
          status: 'error',
          message: String(error),
        }
      }
    }))

    app.use('/', eventHandler((event) => {
      return serveStatic(event, {
        getContents: id => readFile(join(CLIENT_DIR, id)),
        getMeta: async (id) => {
          const stats = await stat(join(CLIENT_DIR, id)).catch(() => {})

          if (!stats || !stats.isFile())
            return

          return {
            size: stats.size,
            mtime: stats.mtimeMs,
          }
        },
      })
    }))

    // @ts-expect-error any
    globalThis.UNPLUGIN_TURBO_CONSOLE_LAUNCH_SERVER = true

    createServer(toNodeListener(app)).listen(port)
  }
}
