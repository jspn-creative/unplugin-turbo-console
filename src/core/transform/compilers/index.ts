import type { CompileResult, Compiler, Context } from '../../../types'
import { vueCompiler } from './vue'
import { vanillaCompiler } from './vanilla'
import { svelteCompiler } from './svelte'

export const compilers: Record<Compiler, (context: Context) => Promise<CompileResult>> = {
  vue: vueCompiler,
  vanilla: vanillaCompiler,
  svelte: svelteCompiler,
}
