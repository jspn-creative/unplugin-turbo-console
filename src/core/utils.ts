import { cwd } from 'node:process'
import { Buffer } from 'node:buffer'
import { extname, relative, sep } from 'pathe'
import { createFilter } from '@rollup/pluginutils'
import type { Comment, Node } from '@babel/types'
import type { Compiler, GenContext } from '../types'
import { PLUGIN_NAME } from './constants'

export const filter = createFilter(
  [/\.vue$/, /\.vue\?vue/, /\.vue\?v=/, /\.ts$/, /\.tsx$/, /\.js$/, /\.jsx$/, /\.svelte$/, /\.astro$/],
  [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/, /[\\/]\.nuxt[\\/]/],
)

const commonStyle = 'padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;'

const consoleStyles: Record<string, string> = {
  '.js': `${commonStyle}color: #111827; background: #F7DF1E`,
  '.jsx': `${commonStyle}color: #111827; background: #F7DF1E`,
  '.ts': `${commonStyle}color: #fff; background: #3178C6`,
  '.tsx': `${commonStyle}color: #fff; background: #3178C6`,
  '.vue': `${commonStyle}color: #fff; background: #4FC08D`,
  '.svelte': `${commonStyle}color: #fff; background: #FF3E00`,
  '.astro': `${commonStyle}color: #fff; background: #FF5D01`,
  'default': `${commonStyle}color: #111827; background: #F7DF1E`,
}

export function getConsoleStyle(fileType: string): string {
  return consoleStyles[fileType] ?? consoleStyles.default
}

export const launchEditorStyle = 'background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px'

export function printInfo(port: number) {
  // eslint-disable-next-line no-console
  console.log('\x1B[32m%s\x1B[0m\x1B[1m%s\x1B[0m\x1B[36m%s\x1B[0m', '  âžœ', `  TurboConsole:`, ` http://localhost:${port}/intro`)
}

export function getFileNameWithoutExtension(fileName: string) {
  if (!fileName)
    return ''
  return fileName.replace(/\.[^/.]+$/, '')
}

export function getExtendedPath(filePath: string, extendedPathFileNames?: string[]) {
  const arr = filePath.split(sep)
  let basename = arr.pop() || ''
  const basenameWithoutExt = getFileNameWithoutExtension(basename).toLowerCase()
  if (extendedPathFileNames && extendedPathFileNames?.length > 0) {
    let isEnd = false
    if (extendedPathFileNames.some(name => basenameWithoutExt === name.toLowerCase())) {
      while (!isEnd) {
        const fileName = arr.pop()
        basename = `${fileName}/${basename}`
        if (extendedPathFileNames.every(name => name.toLowerCase() !== fileName?.toLowerCase()))
          isEnd = true
      }
    }
  }
  return basename
}

export function getCompiler(id: string): Compiler | undefined {
  const urlObject = new URL(id, 'file://')
  const fileType = extname(urlObject.pathname)

  switch (fileType) {
    case '.vue':
      return 'vue'
    case '.svelte':
      return 'svelte'
    case '.js':
      return 'vanilla'
    case '.jsx':
      return 'vanilla'
    case '.ts':
      return 'vanilla'
    case '.tsx':
      return 'vanilla'
    case '.astro':
      return 'vanilla'
    default:
      return undefined
  }
}

export function isPluginDisable(meta: {
  comments: Comment[]
  originalLine: number
  id: string
  type: 'top-file' | 'inline-file'
}) {
  const { comments, originalLine, id, type } = meta

  if (comments?.length === 0)
    return false

  if (type === 'top-file') {
    const compiler = getCompiler(id)
    const startLine = compiler === 'vanilla' ? 1 : 2

    const disablePluginComment = comments?.find(comment => comment.value.includes('turbo-console-disable'))

    if (disablePluginComment && disablePluginComment.loc!.start.line <= startLine)
      return true
  }
  else if (type === 'inline-file') {
    const currentLineComment = comments?.find(comment => comment.value.includes('turbo-console-disable-line') && comment.loc!.start.line === originalLine)
    const nextLineComment = comments?.find(comment => comment.value.includes('turbo-console-disable-next-line') && comment.loc!.start.line === originalLine - 1)

    if (currentLineComment || nextLineComment)
      return true
  }

  return false
}

export function genConsoleString(genContext: GenContext) {
  const { options, originalColumn, originalLine, argType, id } = genContext
  let { argsName } = genContext
  const { prefix, suffix, disableLaunchEditor, port, disableHighlight, extendedPathFileNames } = options
  const _prefix = prefix ? `${prefix}\\n` : ''
  const _suffix = suffix ? `\\n${suffix}` : ''

  const urlObject = new URL(id, 'file://')
  const filePath = urlObject.pathname
  const fileName = getExtendedPath(filePath, extendedPathFileNames)
  const fileType = extname(filePath)

  // Parsing escaped unicode symbols
  try {
    argsName = JSON.parse(`"${argsName}"`)
  }
  catch (error) {
    console.error(`${PLUGIN_NAME}:${error}`)
  }

  if (argsName?.length > 30)
    argsName = `${argsName.slice(0, 30)}...`

  // not output when argtype is string or number
  const lineInfo = `%cðŸš€ ${fileName}\u00B7${originalLine}${['StringLiteral', 'NumericLiteral'].includes(argType) ? '' : ` ~ ${argsName}`}`
  const codePosition = `${relative(cwd(), filePath)}:${originalLine}:${(originalColumn || 0) + 1}`

  const launchEditorString = `%cðŸ”¦ http://localhost:${port}#${Buffer.from(codePosition, 'utf-8').toString('base64')}`

  let consoleString = ''

  if (!disableHighlight && !disableHighlight) {
    consoleString = _prefix
      ? `"${_prefix}${lineInfo}${launchEditorString}","${getConsoleStyle(fileType)}","${launchEditorStyle}","\\n",`
      : `"${lineInfo}${launchEditorString}","${getConsoleStyle(fileType)}","${launchEditorStyle}","\\n",`
  }

  if (disableHighlight && !disableLaunchEditor) {
    consoleString = _prefix
      ? `"${_prefix}${launchEditorString}","${launchEditorStyle}","\\n",`
      : `"${launchEditorString}","${launchEditorStyle}","\\n",`
  }

  if (!disableHighlight && disableLaunchEditor) {
    consoleString = _prefix
      ? `"${_prefix}${lineInfo}","${getConsoleStyle(fileType)}","\\n",`
      : `"${lineInfo}","${getConsoleStyle(fileType)}","\\n",`
  }

  if (disableHighlight && disableLaunchEditor) {
    consoleString = _prefix
      ? `"${_prefix}",`
      : ''
  }

  return {
    consoleString,
    _suffix,
  }
}

export function isConsoleExpression(node: Node) {
  return node.type === 'CallExpression'
    && node.callee.type === 'MemberExpression'
    && node.callee.object.type === 'Identifier'
    && node.callee.object.name === 'console'
    && node.callee.property.type === 'Identifier'
    && node.arguments?.length > 0
}
