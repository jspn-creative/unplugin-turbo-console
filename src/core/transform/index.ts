import type { WithScope } from 'ast-kit'
import { babelParse, walkAST } from 'ast-kit'
import type { Node } from '@babel/types'
import MagicString from 'magic-string'
import type { Compiler, NewContext } from './../../types'
import { compilers } from './compilers'
import { genConsoleString, isConsoleExpression } from './common'

function getCompiler(id: string): Compiler | false {
  const vuePatterns = [/\.vue$/, /\.vue\?vue/, /\.vue\?v=/]
  const sveltePatterns = [/\.svelte$/]
  const jsExpandPatterns = [/\.js$/, /\.jsx$/, /\.ts$/, /\.tsx$/]

  if (vuePatterns.some(pattern => pattern.test(id)))
    return 'vue'
  else if (jsExpandPatterns.some(pattern => pattern.test(id)))
    return 'vanilla'
  else if (sveltePatterns.some(pattern => pattern.test(id)))
    return 'svelte'

  return false
}

export async function transform(context: NewContext) {
  const { code, id, options } = context
  const magicString = new MagicString(code)

  const compiler = getCompiler(id)
  if (!compiler)
    return false

  const compileResult = await compilers[compiler](context)

  const program = babelParse(compileResult.script, 'language', {
    sourceFilename: id,
    plugins: ['jsx', 'typescript'],
  })

  walkAST<WithScope<Node>>(program, {
    enter(node) {
      if (isConsoleExpression(node)) {
        const expressionStart = node.start!
        const expressionEnd = node.end!

        const originalExpression = magicString.slice(expressionStart, expressionEnd)

        if (originalExpression.includes('%c'))
          return false

        const { line, column } = node.loc!.start
        // @ts-expect-error any
        const args = node.arguments

        const argsStart = args[0].start! + compileResult.offset
        const argsEnd = args[args.length - 1].end! + compileResult.offset
        const argType = args[0].type

        const argsName = magicString.slice(argsStart, argsEnd)
          .toString()
          .replace(/`/g, '')
          .replace(/\n/g, '')
          .replace(/"/g, '')

        const originalLine = line + compileResult.line
        const originalColumn = column

        const { consoleString, _suffix } = genConsoleString({
          options,
          originalLine,
          originalColumn,
          argType,
          argsName,
          id,
        })

        consoleString && magicString.appendLeft(argsStart, consoleString)
        _suffix && magicString.appendRight(argsEnd, `,"${_suffix}"`)
      }
    },
  })

  return {
    code: magicString.toString(),
    map: magicString.generateMap({ source: id }),
  }
}
