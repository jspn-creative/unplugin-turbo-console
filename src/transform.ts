import { basename, extname, relative } from 'node:path'
import { cwd } from 'node:process'
import { Buffer } from 'node:buffer'
import type { WithScope } from 'ast-kit'
import { babelParse, getLang, walkAST } from 'ast-kit'
import type { Node } from '@babel/types'
import type { RawSourceMap } from 'source-map'
import { SourceMapConsumer } from 'source-map'
import MagicString from 'magic-string'
import { getConsoleStyle, launchEditorStyle } from './utils/style'
import type { Context } from './types'

export async function transformCode(code: string, id: string, context: Context) {
  try {
    const magicString = new MagicString(code)
    const program = babelParse(code, getLang(id), {
      sourceFilename: id,
    })
    const { port, protocol, base } = context.network

    const asyncOperations: Promise<void>[] = []

    walkAST<WithScope<Node>>(program, {
      enter(node) {
        if (node.type === 'CallExpression'
            && node.callee.type === 'MemberExpression'
            && node.callee.object.type === 'Identifier'
            && node.callee.object.name === 'console'
            && node.callee.property.type === 'Identifier'
            && node.callee.property.name === 'log'
            && node.arguments?.length > 0
        ) {
          const fileName = basename(id)
          const fileType = extname(id)

          const { line, column } = node.loc!.start
          const args = node.arguments

          const sourceMap = context.pluginContext.getCombinedSourcemap()
          const asyncOperation = new SourceMapConsumer(sourceMap as RawSourceMap).then((consumer) => {
            const { line: originalLine, column: originalColumn } = consumer.originalPositionFor({
              line,
              column,
            })

            const argsStart = args[0].start!
            const argsEnd = args[args.length - 1].end!
            const argType = args[0].type
            const { prefix, suffix, disableLaunchEditor } = context.option
            const _prefix = prefix ? `${prefix} \\n` : ''
            const _suffix = suffix ? `\\n ${suffix}` : ''

            const argsName = magicString.slice(argsStart, argsEnd)
              .toString()
              .replace(/`/g, '')
              .replace(/\n/g, '')
              .replace(/"/g, '')

            // not output when argtype is string or number
            const lineInfo = `${_prefix}%cðŸš€ ${fileName}:${originalLine}${['StringLiteral', 'NumericLiteral'].includes(argType) ? '' : ` ~ ${argsName}`}`
            const codePosition = `${relative(cwd(), id)}:${originalLine}:${(originalColumn || 0) + 1}`

            let launchEditorString = ''
            launchEditorString = `%cðŸ”¦ Jump to Editor ${protocol}://localhost:${port}${base}__tc/i.html`

            if (base !== '/')
              launchEditorString += `?b=${base}`

            launchEditorString += `#${Buffer.from(codePosition, 'utf-8').toString('base64')}`

            let appendLeftString = ''

            if (!disableLaunchEditor)
              appendLeftString = `"${lineInfo} %c\\n${launchEditorString}","${getConsoleStyle(fileType)}","","${launchEditorStyle}","\\n",`

            else
              appendLeftString = `"${lineInfo} %c\\n","${getConsoleStyle(fileType)}","\\n",`

            magicString
              .appendLeft(argsStart, appendLeftString)
              .appendRight(argsEnd, `,"${_suffix}"`)
          })

          asyncOperations.push(asyncOperation)
        }
      },
    })

    await Promise.all(asyncOperations)

    return {
      code: magicString.toString(),
      map: magicString.generateMap({ source: id, includeContent: true }),
    }
  }
  catch (error) {
    // eslint-disable-next-line no-console
    console.log(error)
    throw new Error(String(error))
  }
}
