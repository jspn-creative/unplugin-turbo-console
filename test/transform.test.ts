import { join } from 'pathe'
import { describe, expect, it, vi } from 'vitest'
import { transform } from '../src/core/transform'
import { resolveOptions } from '../src/core/options'
import { INCLUDES_HIGHLIGHT, SVELTE, TSX, TYPESCRIPT, VUE_OPTIONS, VUE_SCRIPT_SETUP } from './fixtures/transform'

vi.mock('node:process', () => {
  return {
    cwd: vi.fn(() => join('/', 'mock', 'path')),
    env: {
      NODE_ENV: 'development',
    },
  }
})

const mockRouteMap = new Map()
mockRouteMap.set('../../home/runner/App.vue', 'appvue')
mockRouteMap.set('../../home/runner/main.ts', 'maints')
mockRouteMap.set('../../home/runner/main.js', 'mainjs')
mockRouteMap.set('../../home/runner/+page.svelte', 'pagtsx')
mockRouteMap.set('../../home/runner/page.tsx', 'svelte')

globalThis.TurboConsoleRouteMap = mockRouteMap

describe('vue transform', () => {
  it ('script setup', async () => {
    VUE_SCRIPT_SETUP.options = resolveOptions(VUE_SCRIPT_SETUP.options)
    expect(await transform(VUE_SCRIPT_SETUP)).matchSnapshot()
  })

  it('options', async () => {
    VUE_OPTIONS.options = resolveOptions(VUE_OPTIONS.options)
    expect(await transform(VUE_OPTIONS)).matchSnapshot()
  })
})

describe('vanilla transform', () => {
  it('typescript', async () => {
    TYPESCRIPT.options = resolveOptions(TYPESCRIPT.options)
    expect(await transform(TYPESCRIPT)).matchSnapshot()
  })

  it('tsx', async () => {
    TSX.options = resolveOptions(TSX.options)
    expect(await transform(TSX)).matchSnapshot()
  })
})

describe('svelte transform', () => {
  it('svelte', async () => {
    SVELTE.options = resolveOptions(SVELTE.options)
    expect(await transform(SVELTE)).matchSnapshot()
  })
})

describe('edge case', () => {
  it('includes highlight', async () => {
    INCLUDES_HIGHLIGHT.options = resolveOptions(INCLUDES_HIGHLIGHT.options)
    expect(await transform(INCLUDES_HIGHLIGHT)).matchSnapshot()
  })
})
