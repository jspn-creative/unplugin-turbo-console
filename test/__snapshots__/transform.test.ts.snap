// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`edge case > includes highlight 1`] = `
{
  "code": "console.log('%cLog Message', 'color: orange');",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA",
    "names": [],
    "sources": [
      "/home/runner/src/main.js",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`svelte transform > svelte 1`] = `
{
  "code": "<script lang="ts">
  // xxxx
  const a: string = 'Hello'
  console.log("%c🚀 +page.svelte·4 ~ a%c🔦 http://localhost:3070#Li4vLi4vK3BhZ2Uuc3ZlbHRlOjQ6Mw==","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #FF3E00","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px","\\n",a)
</script>

<h1>Welcome to SvelteKit</h1>
<p>
  Visit <a href="https://kit.svelte.dev">kit.svelte.dev</a> to read the documentation
</p>",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA,4RAAc;AACd;AACA;AACA;AACA;AACA;AACA",
    "names": [],
    "sources": [
      "+page.svelte",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`vanilla transform > tsx 1`] = `
{
  "code": ""use client";

  export default function Home() {
    console.log("%c🚀 page.tsx·4%c🔦 http://localhost:3070#Li4vLi4vcGFnZS50c3g6NDo1","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #3178C6","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px","\\n","hello nextjs");
    function hello() {
      console.log("%c🚀 page.tsx·6%c🔦 http://localhost:3070#Li4vLi4vcGFnZS50c3g6Njo3","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #3178C6","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px","\\n","hello nextjs");
    }
  
    return <main>page.tsx</main>;
  }
  ",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA,8QAAgB;AAChB;AACA,gRAAkB;AAClB;AACA;AACA;AACA;AACA",
    "names": [],
    "sources": [
      "page.tsx",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`vanilla transform > typescript 1`] = `
{
  "code": "const msg: string = "hello TypeScript"
  console.log("%c🚀 main.ts·2 ~ msg%c🔦 http://localhost:3070#Li4vLi4vbWFpbi50czoyOjM=","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #3178C6","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px","\\n",msg)
  ",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,iRAAc;AACd",
    "names": [],
    "sources": [
      "main.ts",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`vue transform > options 1`] = `
{
  "code": "<script>
  export default {
    data() {
      return {
        greeting: 'Hello World!'
      }
    },
    mounted() {
      const msg = "hello vue"
      console.log("%c🚀 App.vue·10 ~ msg%c🔦 http://localhost:3070#Li4vLi4vQXBwLnZ1ZToxMDo3","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #4FC08D","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px","\\n",msg)
    }
  }
  </script>
  
  <template>
    <p class="greeting">{{ greeting }}</p>
  </template>
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA,sRAAkB;AAClB;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`vue transform > script setup 1`] = `
{
  "code": "<template>
  Hello World
</template>

<script setup lang="ts">
  const msg: string = "Hello World"
  console.log("%c🚀 App.vue·7 ~ msg%c🔦 http://localhost:3070#Li4vLi4vQXBwLnZ1ZTo3OjM=","padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;color: #fff; background: #4FC08D","background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px","\\n",msg)
</script>
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA,iRAAc;AACd;",
    "names": [],
    "sources": [
      "App.vue",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
